internal/provider/resource_gitlab_personal_access_tokens.go

```diff
@@ -0,0 +1,242 @@
+				Type:        schema.TypeBool,
+				Computed:    true,
+			},
+			"created_at": {
+				Description: "Time the token has been created, RFC3339 format.",
+				Type:        schema.TypeString,
+				Computed:    true,
+			},
+			"expires_at": {
+				Description:      "The token expires at midnight UTC on that date. The date must be in the format YYYY-MM-DD. Default is never.",
+				Type:             schema.TypeString,
+				Optional:         true,
+				ForceNew:         true,
+				ValidateDiagFunc: isISO6801Date,
+			},
+			"token": {
+				Description: "The personal access token. This is only populated when creating a new personal access token. This attribute is not available for imported resources.",
+				Type:        schema.TypeString,
+				Computed:    true,
+				Sensitive:   true,
+			},
+		},
+	}
+})
+
+func resourceGitlabPersonalAccessTokenCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*gitlab.Client)
+
+	userId := d.Get("user_id").(int)
+
+	currentUserAdmin, err := isCurrentUserAdmin(client)
+	if err != nil {
+		return diag.Errorf("[ERROR] cannot query the user API for current user: %v", err)
+	}
+
+	if !currentUserAdmin {
+		return diag.Errorf("[ERROR] Current user needs to be admin when creating a personal access token: %v", err)
+	}
+
+	options := &gitlab.CreatePersonalAccessTokenOptions{
+		Name:   gitlab.String(d.Get("name").(string)),
+		Scopes: stringSetToStringSlice(d.Get("scopes").(*schema.Set)),
+	}
+
+	log.Printf("[DEBUG] create gitlab PersonalAccessToken %s (scopes: %s) for user ID %d", *options.Name, options.Scopes, userId)
+
+	if v, ok := d.GetOk("expires_at"); ok {
+		parsedExpiresAt, err := time.Parse("2006-01-02", v.(string))
+		if err != nil {
+			return diag.Errorf("Invalid expires_at date: %v", err)
+		}
+
+		parsedExpiresAtISOTime := gitlab.ISOTime(parsedExpiresAt)
+		options.ExpiresAt = &parsedExpiresAtISOTime
+	}
+
+	personalAccessToken, _, err := client.Users.CreatePersonalAccessToken(userId, options, gitlab.WithContext(ctx))
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	tokenId := strconv.Itoa(personalAccessToken.ID)
+	userIdString := strconv.Itoa(userId)
+
+	d.SetId(buildTwoPartID(&userIdString, &tokenId))
+	// NOTE: the token can only be read once after creating it
+	d.Set("token", personalAccessToken.Token)
+
+	return resourceGitlabPersonalAccessTokenRead(ctx, d, meta)
+}
+
+func resourceGitlabPersonalAccessTokenRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	userId, tokenId, err := parseTwoPartID(d.Id())
+	if err != nil {
+		return diag.Errorf("Error parsing ID: %s", d.Id())
+	}
+
+	userIdInt, err := strconv.Atoi(userId)
+	if err != nil {
+		return diag.Errorf("Error converting user ID to string: %v", userId)
+	}
+
+	client := meta.(*gitlab.Client)
+
+	personalAccessTokenId, err := strconv.Atoi(tokenId)
+	if err != nil {
+		return diag.Errorf("%s cannot be converted to int", tokenId)
+	}
```

Comment:
Can we refactor this into it's own function, like here:

https://github.com/gitlabhq/terraform-provider-gitlab/blob/4337c7437f809c2190d3e07e94137fa58b734729/internal/provider/resource_gitlab_project_issue.go#L240